### PR TITLE
fix: respect environment during `bentoml.bentos.build`

### LINKED ISSUE #4080: bug: Can't modify BENTOML_HOME when running build_bentofile from Python
### Describe the bug

Changes to the BentoML environment configuration made via the Python APIs don't get propagated to the `build` workflow because the build now runs in a subprocess in `v1.1`.

### To reproduce

I have a Python app that's sort of a Bento "build server", so it needs to build Bentos in temporary directories. In BentoML v1.0, I used the following code to accomplish this:

```
with tempfile.TemporaryDirectory() as tmp, \
     bentoml.bentos.BentoMLContainer.bentoml_home.patch(tmp):
    ... create the model
    bentoml.<framework>.save_model(...)
    ... create bentofile.yaml under tmp
    bentoml.bentos.build_bentofile(build_ctx=tmp)
```

In v1.1, the implementation of `build_bentofile` has changed to delegate to the CLI app running in a subprocess. As a result, it has a different copy of the configuration objects, and always uses whatever is set for `BENTOML_HOME` at the system level. This results in an error like:

```
Error: [bentoml-cli] `build` failed: no Models with name <name> exist in BentoML store <osfs '/root/bentoml/models'>
```

because `BENTOML_HOME` was (the value of) `tmp` when `save_model()` was called, but in the build it has reverted to its default value of `~/bentoml`.

### Expected behavior

The build should run as though `BENTOML_HOME` had been set to the temporary directory, i.e., the equivalent of `BENTOML_HOME=<value of tmp> bentoml build ...`

### Environment

bentoml: 1.1
python: 3.8
Linux in a Docker container

### PR DESCRIPTION
**Pull request by aarnphm**
## What does this PR address?

Fixes #4080

Passing environment into subprocess build

Signed-off-by: Aaron <29749331+aarnphm@users.noreply.github.com>

## Before submitting:

- [ ] Does the Pull Request follow [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/#summary) naming? Here are [GitHub's
      guide](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request) on how to create a pull request.
- [ ] Does the code follow BentoML's code style, `pre-commit run -a` script has passed ([instructions](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#style-check-auto-formatting-type-checking))?
- [ ] Did you read through [contribution guidelines](https://github.com/bentoml/BentoML/blob/main/CONTRIBUTING.md#ways-to-contribute) and follow [development guidelines](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#start-developing)?
- [ ] Did your changes require updates to the documentation? Have you updated
      those accordingly? Here are [documentation guidelines](https://github.com/bentoml/BentoML/tree/main/docs) and [tips on writting docs](https://github.com/bentoml/BentoML/tree/main/docs#writing-documentation).
- [ ] Did you write tests to cover your changes?

### CODE REVIEW

File: src/bentoml/bentos.py [hunk 1]
```diff
@@ -369,23 +369,38 @@ def build(
         build_args.extend(["--version", version])
     build_args.extend(["--output", "tag"])
 
+    copied = os.environ.copy()
+    if "BENTOML_HOME" not in copied:
+        copied["BENTOML_HOME"] = BentoMLContainer.bentoml_home.get()
```
Comment (frostming) [inline src/bentoml/bentos.py:374]:
```suggestion
    copied.setdefault("BENTOML_HOME", BentoMLContainer.bentoml_home.get())
```

File: src/bentoml/bentos.py [hunk 2]
```diff
@@ -421,17 +436,18 @@ def build_bentofile(
         build_args.extend(["--version", version])
     build_args.extend(["--bentofile", bentofile, "--output", "tag"])
 
+    copied = os.environ.copy()
+    if "BENTOML_HOME" not in copied:
+        copied["BENTOML_HOME"] = BentoMLContainer.bentoml_home.get()
```
Comment (frostming) [inline src/bentoml/bentos.py:441]:
```suggestion
    copied.setdefault("BENTOML_HOME", BentoMLContainer.bentoml_home.get())
```

File: src/bentoml/bentos.py [hunk 3]
```diff
@@ -369,23 +369,37 @@ def build(
...
-    _, _, tag = matched.group(0).partition(":")
-    return get(tag, _bento_store=_bento_store)
+
+def _parse_tag_from_outputs(output: bytes) -> str:
+    matched = re.search(
+        r"^__tag__:[^:\n]+:[^:\n]+", output.decode("utf-8").strip(), re.MULTILINE
+    )
+    if matched is None:
+        raise BentoMLException(
+            f"Failed to find tag from output: {output}\nNote: Output from 'bentoml build' might not be correct. Please open a tickets on GitHub."
```
Comment (frostming) [inline src/bentoml/bentos.py:400]:
```suggestion
            f"Failed to find tag from output: {output}\nNote: Output from 'bentoml build' might not be correct. Please open an issue on GitHub."
```

File: src/bentoml/bentos.py [hunk 4]
```diff
@@ -369,23 +369,37 @@ def build(
...
 
-    pattern = r"^__tag__:[^:\n]+:[^:\n]+"
-    matched = re.search(pattern, output.decode("utf-8").strip(), re.MULTILINE)
-    assert matched is not None, f"Failed to find tag from output: {output}"
-    _, _, tag = matched.group(0).partition(":")
-    return get(tag, _bento_store=_bento_store)
+
+def _parse_tag_from_outputs(output: bytes) -> str:
+    matched = re.search(
+        r"^__tag__:[^:\n]+:[^:\n]+", output.decode("utf-8").strip(), re.MULTILINE
```
Comment (frostming) [inline src/bentoml/bentos.py:396]:
```suggestion
    matched = re.match(
        r"__tag__:([^:\n]+:[^:\n]+)$", output.decode("utf-8").strip()
```

File: src/bentoml/bentos.py [hunk 5]
```diff
@@ -369,23 +369,37 @@ def build(
...
+    )
+    if matched is None:
+        raise BentoMLException(
+            f"Failed to find tag from output: {output}\nNote: Output from 'bentoml build' might not be correct. Please open a tickets on GitHub."
+        )
+    return matched.group(0).partition(":")[-1]
```
Comment (frostming) [inline src/bentoml/bentos.py:402]:
```suggestion
    return matched.group(1)
```

Comment (aarnphm) [reply src/bentoml/bentos.py:402]: 😄

**Review by frostming — DISMISSED** [review]

**Comment by aarnphm** [comment]
cc @frostming I just ran the pre-commit.

**Review by frostming — APPROVED** [review]

**Pull request merged**